doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md

```diff
@@ -0,0 +1,1069 @@
+      med 10
+    Set clauses:
+    Call clauses:
+    Actions:
+      Exit routemap
+sonic#
+```
+|Command Description|CLI Command      |
+|:------------------|:-----------------|
+| Display IPv4 prefix list information | show ip prefix-list |
+
+```
+sonic# show ip prefix-list
+IP prefix list pref1:
+     permit 20.0.0.0/8
+sonic#
+```
+|Command Description|CLI Command      |
+|:------------------|:-----------------|
+| Display IPv6 prefix list information | show ipv6 prefix-list |
+
+```
+sonic# show ipv6 prefix-list
+IPv6 prefix list pref2:
+     permit 2222::/64 ge 65 le 65
+     permit 2223::/64 ge 65 le 128
+sonic#
+```
+|Command Description|CLI Command      |
+|:------------------|:-----------------|
+| Display BGP community list information | show bgp community-list |
+
+```
+sonic# show bgp community-list
+Standard community list com1:  match: ANY
+     local-AS
+Expanded community list com2:   match: ANY
+     Extended1
+sonic#
+```
+|Command Description|CLI Command      |
+|:------------------|:-----------------|
+| Display BGP extended community list information | show bgp ext-community-list |
+
+```
+sonic# show bgp ext-community-list
+Standard extended community list extcom1:  match: ANY
+     rt:2:2
+Expanded extended community list extcom2:  match: ANY
+     extcom
+sonic#
+```
+
+|Command Description|CLI Command      |
+|:------------------|:-----------------|
+| Display BGP AS-Path list information | show bgp as-path-access-list |
+
+```
+sonic# show bgp as-path-access-list
+AS path list aspath1:
+   members: 1:1+
+sonic#
+```
+
+
+#### 3.6.2.3 Debug Commands
+N/A
+
+#### 3.6.2.4 IS-CLI Compliance
+NA
+### 3.6.3 REST API Support
+Various REST operations (POST/PUT/PATCH/GET/DELETE) are supported for the BGP and route map configurations.
+# 4 Flow Diagrams
+
+## 4.1 Configuration Sequence
+
+![FRR-BGP-CONFIG-SEQUENCE](images/frr-bgp-config-sequence.jpg)
+
+## 4.2 CLI Show Command Sequence
+
+### 4.2.1 CLI Show Sequence - Config information
+![FRR-BGP-CLI-SHOW-SEQUENCE1](images/frr-bgp-cli-show-sequence11.jpg)
+
+### 4.2.2 CLI Show Sequence - State/Statistics
+![FRR-BGP-CLI-SHOW-SEQUENCE2](images/frr-bgp-cli-show-sequence22.jpg)
+
+## 4.3 REST Get Sequence
+
+### 4.3.1 REST Get Sequence - Config information
+![FRR-BGP-REST-GET-SEQUENCE1](images/frr-bgp-rest-get-sequence1.jpg)
+
+### 4.3.2 REST Get Sequence - State/Statistics
+![FRR-BGP-REST-GET-SEQUENCE2](images/frr-bgp-rest-get-sequence2.jpg)
+
+# 5 Error Handling
+
+The complete validation of BGP parameters is performed in the SONiC management framework based on open config BGP/Routing policy YANGs and against corresponding SONiC YANGs. Appropriate error code is returned for invalid configurations or on failures due to a dependency from management framework and bgpcfgd modules and the same is logged into the syslog, it is not expected for bgpcfgd (FRR CLI command execution) to return validation errors. However, such errors are logged into syslog for debugging purposes.
```

Comment:
1. It's impossible to validate frr configuration using only YANG models. There is no semantics in the models
2. what if bgpcfgd reports any errors - rest api will show no errors. I think we must show errors from bgpcfgd in the api

Reply:
For now, we use syslog to report an error for invalid configs, looks like currently we dont report any BGP config error to the user from bgpcfgd, do you have any suggestion?

Reply:
My opinion is to create working configuration for bgpd. syslog report will not work for me. How big is a chance that someone with REST API will check syslog output on the box during configuration? We must give correct output to users.

Reply:
With the help of config cache in bgpcfgd, we resolve the config dependency issues, but other than that all other validations are expected to be done in mgmt-framework backend code, I dont see any reason for throwing an error to syslog from bgpcfgd for valid configs, if there are any negative configs and we should throw error for those cases, we can handle later as and when we encounter them..what do you say?

Reply:
In your design bgpcfgd is not connected to your frontends.
bgpcfgd reads the database only. Your design doesn't have any feedback mechanism from bgpcfgd to the frontends. I don't agree with your solution to output issues into the syslog from bgpcfgd. That means that the frontends will never see the issues.

Reply:
I understand that the current design does not have any feedback mechanism, will explore options for notifying the errors to user (front-end) in the future.

Reply:
As a note, SONiC in general does not provide a means to notify front end users of back end errors (this not just an issue of this feature). It seems that there is work underway (another activity) to provide a means for back-end applications to notify front-end applications when an error occurs at the backend. We'll look into using that new framework when it becomes available.

My understanding is that the overall SONiC design, both for integrating 3rd party applications (e.g. DHCP and teamd etc.), and communication between its native components, uses the following approach:
-	A front end application (e.g. SONiC CLI/CLICK) or some other application performs configuration validation (semantic and syntactic)
-	If the configuration is valid, it is written to the Redis DB, which publishes DB change events
-	A back-end applications listens to those DB change events and pushes the changes to the 3rd party app (DHCP, teamd etc.) or a native SONiC component. 
o	Validation needs to happen at the front end – by the time the configuration data reached the back end, it has already been written to the DB

In the context of this feature, the SONiC Mgmt Framework (BGP REST and CLI implementation) is the front-end application, and bgpcfgd is the backend application. 
o	If we wrote data into FRR first (the only means to fully validate BGP config) – before writing data to the Redis DB – we’d go against the overall concepts of the SONiC architecture.